tsl/src/compression/api.c

```diff
@@ -892,3 +976,655 @@ tsl_recompress_chunk_wrapper(Chunk *uncompressed_chunk)
+	Chunk *compressed_chunk = ts_chunk_get_by_id(uncompressed_chunk->fd.compressed_chunk_id, true);
+
+	List *htcols_list = ts_hypertable_compression_get(srcht_id);
+	ListCell *lc;
+	int htcols_listlen = list_length(htcols_list);
+
+	const ColumnCompressionInfo **colinfo_array;
+	colinfo_array = palloc(sizeof(ColumnCompressionInfo *) * htcols_listlen);
+
+	int nsegmentby_cols = 0, i = 0;
+
+	foreach (lc, htcols_list)
+	{
+		FormData_hypertable_compression *fd = (FormData_hypertable_compression *) lfirst(lc);
+		colinfo_array[i++] = fd;
+		if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			nsegmentby_cols++;
+	}
+
+	const ColumnCompressionInfo **keys;
+	int n_keys;
+	int16 *in_column_offsets = compress_chunk_populate_keys(uncompressed_chunk->table_id,
+															colinfo_array,
+															htcols_listlen,
+															&n_keys,
+															&keys);
+
+	Relation compressed_chunk_rel = table_open(compressed_chunk->table_id, AccessShareLock);
+	Relation uncompressed_chunk_rel = table_open(uncompressed_chunk->table_id, AccessShareLock);
+	TupleDesc compressed_rel_tupdesc = RelationGetDescr(compressed_chunk_rel);
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+
+	RowCompressor row_compressor;
+	row_compressor_init(&row_compressor,
+						uncompressed_rel_tupdesc,
+						compressed_chunk_rel,
+						htcols_listlen,
+						colinfo_array,
+						in_column_offsets,
+						compressed_rel_tupdesc->natts,
+						true /*need_bistate*/,
+						true /*recompressing*/);
+
+	if (OidIsValid(row_compressor.index_oid))
+	{
+		/* lock in least strict mode so it's not dropped by other sessions */
+		Relation index_rel = relation_open(row_compressor.index_oid, AccessExclusiveLock);
+		relation_close(index_rel, NoLock);
+	}
+
+	table_close(compressed_chunk_rel, AccessShareLock);
+	table_close(uncompressed_chunk_rel, AccessShareLock);
+
+	row_compressor_finish(&row_compressor);
+
+	if (OidIsValid(row_compressor.index_oid))
+	{
+		PG_RETURN_OID(uncompressed_chunk_id);
+	}
+	else
+		PG_RETURN_NULL();
+}
+
+/*
+ * This function fetches the remaining uncompressed chunk rows into
+ * the tuplestore for recompression.
+ * `unmatched` here means that no existing segment in the compressed
+ * chunk matches the segmentby column values for these remaining rows,
+ * so new segments must be created for them.
+ */
+static void
+fetch_unmatched_uncompressed_chunk_into_tuplestore(Tuplesortstate *segment_tuplesortstate,
+												   Relation uncompressed_chunk_rel)
+{
+	TableScanDesc heapScan;
+	HeapTuple uncompressed_tuple;
+	TupleDesc uncompressed_rel_tupdesc = RelationGetDescr(uncompressed_chunk_rel);
+
+	heapScan = table_beginscan(uncompressed_chunk_rel, GetLatestSnapshot(), 0, NULL);
+
+	for (uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
+		 uncompressed_tuple != NULL;
+		 uncompressed_tuple = heap_getnext(heapScan, ForwardScanDirection))
+	{
+		TupleTableSlot *heap_tuple_slot =
```

Comment:
Couldn't this `heap_tuple_slot ` be cleared and reused in the entire loop? This could improve performance.